### PR TITLE
fixed bunch of memory leaks in json constructor

### PR DESCRIPTION
**Pull request by dpronin**
Launch fio built with ASAN enabled, for example, by this way:
```
./fio --max-jobs=4 --output=fio-output.json --output-format=json+ --filename=fio_jsonplus_clat2csv.test --ioengine=libaio --time_based --runtime=3s --size=1M --slat_percentiles=1 --clat_percentiles=1 --lat_percentiles=1 --thread=1 --name=test1 --rw=randrw --name=test2 --rw=read --name=test3 --rw=write --thread=1
```

Borrowed from one tests the project has got

We can see many memory leaks given by JSON filler/parser:

```
Direct leak of 49 byte(s) in 9 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3ee62f in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3ee62f in json_add_job_opts /home/dpronin/Development/workspace/fio/stat.c:1674:3
    #3 0x3ee62f in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2648:3
    #4 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #5 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9

Direct leak of 38 byte(s) in 7 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3c648c in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3c648c in json_add_job_opts /home/dpronin/Development/workspace/fio/stat.c:1674:3
    #3 0x3c648c in show_thread_status_json /home/dpronin/Development/workspace/fio/stat.c:1706:3
    #4 0x3ee912 in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2667:31
    #5 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #6 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9

Direct leak of 25 byte(s) in 1 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3ee4aa in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3ee4aa in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2646:3
    #3 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #4 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9

Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3ee385 in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3ee385 in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2643:3
    #3 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #4 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9

Direct leak of 18 byte(s) in 3 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3c60c0 in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3c60c0 in show_thread_status_json /home/dpronin/Development/workspace/fio/stat.c:1693:2
    #3 0x3ee912 in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2667:31
    #4 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #5 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9

Direct leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x3319c4 in strdup (/home/dpronin/Development/workspace/fio/fio+0x3319c4)
    #1 0x3c2efb in json_object_add_value_string /home/dpronin/Development/workspace/fio/./json.h:85:15
    #2 0x3c2efb in json_array_add_disk_util /home/dpronin/Development/workspace/fio/stat.c:1078:2
    #3 0x3c3bcf in json_object_add_disk_utils /home/dpronin/Development/workspace/fio/stat.c:1124:3
    #4 0x3c3bcf in show_disk_util /home/dpronin/Development/workspace/fio/stat.c:1150:3
    #5 0x3eea8f in __show_run_stats /home/dpronin/Development/workspace/fio/stat.c:2676:3
    #6 0x49cf29 in fio_backend /home/dpronin/Development/workspace/fio/backend.c:2590:3
    #7 0x51dae4 in main /home/dpronin/Development/workspace/fio/fio.c:60:9
```

This PR fixes these leaks

### CODE REVIEW

**Comment by dpronin** [comment]
How do I poke CI one more time?

**Comment by ammarfaizi2** [comment]
> How do I poke CI one more time?

Force push may do?

`git commit --amend` with `--no-edit` will change the commit sha1 hash, so you can force push.
```
git commit --amend --no-edit;
git push --force;
```

File: json.h [hunk 1]
```diff
@@ -81,9 +84,14 @@ static inline int json_object_add_value_string(struct json_object *obj,
...
+	char *string;
 
-	arg.string = strdup(val ? : "");
-	return json_object_add_value_type(obj, name, &arg);
+	string = arg.string = strdup(val ? : "");
+	res = json_object_add_value_type(obj, name, &arg);
+	free(string);
+
+	return res;
 }
```
Comment (ammarfaizi2) [inline json.h:95]: A question...

If the lifetime of `string` is just as short as the `json_object_add_value_type` call, why do we even need `strdup` here? Can't we just do this:
```
arg.string = val ? : "";
return json_object_add_value_type(obj, name, &arg);
```

Comment (ammarfaizi2) [reply json.h:95]: Does `json_object_add_value_type` write to `string[n]`?

Comment (ammarfaizi2) [reply json.h:95]: > Does `json_object_add_value_type` write to `string[n]`?

If not, maybe cast away the const?

**Comment by axboe** [comment]
This looks very wrong. You'd want to free the memory once the json is printed, doing it after adding the value makes zero sense. At that point you might as well just assign the pointer directly.

**Comment by dpronin** [comment]
@axboe why have you closed the PR then? Will we just need to assign the pointer without duplicating? OK, it can be done fast, there was no need to close the PR, was there?

**Comment by dpronin** [comment]
@ammarfaizi2 strdup wasn't introduced by me, I have just got rid of memory leak. I have had no idea why `strdup` was introduced

**Comment by axboe** [comment]
Just send a new PR if it's appropriate.

**Comment by dpronin** [comment]
@axboe we cannot assign the pointer directly because `struct json_value` has `string` as a mutable string whereas we have immutable as input. That is why `strdup` has been used

**Comment by ammarfaizi2** [comment]
> @axboe we cannot assign the pointer directly because `struct json_value` has `string` as a mutable string whereas we have immutable as input. That is why `strdup` has been used

So it seems `arg.string` is just read-only in
this context. If that's the case, I think it
should be fine to cast away the const. Yes? No?

I am not sure what `json_object_add_value_type`
actually does w.r.t. `arg.string`.

**Comment by dpronin** [comment]
@ammarfaizi2 don't you think it is illegal from the call site point of view? casting away 'const' is a cheating and I do not consider this as a good practice. I would suggest we turn mutable string to immutable in `struct json_value`

**Comment by ammarfaizi2** [comment]
> I would suggest we turn mutable string into immutable in `struct json_value`

Well, that breaks here: https://github.com/axboe/fio/pull/1387

Moving the discussion to that PR.

**Pull request closed without merging**